src/fomo.jl

```diff
@@ -0,0 +1,427 @@
+      @info @sprintf "%5d  %9.2e  %7.1e  %7.1e" stats.iter stats.objective norm_∇fk αk
+    end
+    
+  end
+  if verbose > 0 && mod(stats.iter, verbose) == 0
+    if r2mode
+      @info @sprintf "%5s  %9s  %7s  %7s" "iter" "f" "‖∇f‖" "σ"
+    infoline = @sprintf "%5d  %9.2e  %7.1e  %7.1e" stats.iter stats.objective norm_∇fk 1/αk
+    else
+      @info @sprintf "%5s  %9s  %7s  %7s  %7s" "iter" "f" "‖∇f‖" "α" "staβ"
+    infoline = @sprintf "%5d  %9.2e  %7.1e  %7.1e  %7.1e" stats.iter stats.objective norm_∇fk αk 0
+    end
+    
+  end
+
+  set_status!(
+    stats,
+    get_status(
+      nlp,
+      elapsed_time = stats.elapsed_time,
+      optimal = optimal,
+      max_eval = max_eval,
+      iter = stats.iter,
+      max_iter = max_iter,
+      max_time = max_time,
+    ),
+  )
+
+  callback(nlp, solver, stats)
+
+  done = stats.status != :unknown
+
+  d .= ∇fk
+  norm_d = norm_∇fk
+  satβ = T(0)
+  ρk = T(0)
+  avgsatβ = T(0)
+  siter = 0
+  oneT = T(1)
+  while !done
+    λk = step_mult(αk,norm_d,backend)
+    c .= x .- λk .* d
+    uf = x == c # step addition underfow on every dimensions, should happen before αk == 0
+    if r2mode
+      ΔTk = norm_∇fk^2 * λk
+    else  
+      ΔTk = dot(∇fk , d) * λk
+    end
+    fck = obj(nlp, c)
+    if fck == -Inf
+      set_status!(stats, :unbounded)
+      break
+    end
+    ρk = (stats.objective - fck) / ΔTk
+    # Update regularization parameters
+    if ρk >= η2
+      αk = min(αmax, γ2 * αk)
+    elseif ρk < η1
+      αk = αk * γ1
+      if !r2mode
+        satβ *= γ3
+        d .= ∇fk .* (oneT - satβ) .+ m .* satβ
+      end
+    end
+
+    # Acceptance of the new candidate
+    if ρk >= η1
+      x .= c
+      if !r2mode
+        m .= ∇fk .* (oneT - β) .+ m .* β
+      end
+      set_objective!(stats, fck)
+      grad!(nlp, x, ∇fk)
+      norm_∇fk = norm(∇fk)
+      if !r2mode
+        satβ = find_beta(m, ∇fk, norm_∇fk, β, θ1, θ2)
+        d .= ∇fk .* (oneT - satβ) .+ m .* satβ
+        norm_d = norm(d)
+      else
+        d .= ∇fk
+        norm_d = norm_∇fk
+      end
+      if !r2mode
+        avgsatβ += satβ
+        siter += 1
+      end
+    end
+
+    set_iter!(stats, stats.iter + 1)
+    set_time!(stats, time() - start_time)
+    set_dual_residual!(stats, norm_∇fk)
+    optimal = norm_∇fk ≤ ϵ
+
+    if verbose > 0 && mod(stats.iter, verbose) == 0
+      @info infoline
+      if r2mode
+        infoline = @sprintf "%5d  %9.2e  %7.1e  %7.1e" stats.iter stats.objective norm_∇fk 1/αk
+      else
+        infoline = @sprintf "%5d  %9.2e  %7.1e  %7.1e  %7.1e" stats.iter stats.objective norm_∇fk αk satβ
+      end
+    end
+
+    set_status!(
+      stats,
+      get_status(
+        nlp,
+        elapsed_time = stats.elapsed_time,
+        optimal = optimal,
+        max_eval = max_eval,
+        iter = stats.iter,
+        max_iter = max_iter,
+        max_time = max_time,
+      ),
+    )
+    
+    callback(nlp, solver, stats)
+
+    uf      && set_status!(stats,:small_step)
```

Comment:
I couldn’t find where `uf` is defined and have no idea what it means. Try to use self-explanatory variable names.

Reply:
`uf` stands for step underflow, and is true if `x == c`. It is computed at line 295. I'll rename the variable.